File: back-end/src/main/java/es/ull/project/domain/valueobject/File.java
package es.ull.project.domain.valueobject;

import java.net.URI;
import java.util.Optional;
import es.ull.project.domain.enumerate.NodeContentType;

/**
 * @brief Value object representing a file with a name and optional URI.
 */
public class File extends NodeContent {

    private static final String ERROR_FILENAME_NOT_DEFINED = "File name not defined";
    /**
     * Name of the file. It is a required attribute.
     */
    private final FileName name;
    /**
     * URI of the file. It is an optional attribute.
     */
    private final URI uri;

    public File(FileName name, URI uri) {
        validateName(name);
        this.name = name;
        this.uri = uri;
    }

    public File(FileName name) {
        this(name, null);
    }

    private void validateName(FileName name) {
        if (name == null) {
            throw new IllegalArgumentException(ERROR_FILENAME_NOT_DEFINED);
        }
    }

    public FileName getName() {
        return this.name;
    }

    /**
     * @return Optional containing the URI if present, or empty otherwise.
     */
    public Optional<URI> getUri() {
        return Optional.ofNullable(this.uri);
    }

    /**
     * @return true if the file has a URI.
     */
    public boolean hasUri() {
        return uri != null;
    }

    public File setName(FileName name) {
        return new File(name, this.uri);
    }

    public File setUri(URI uri) {
        return new File(this.name, uri);
    }

    @Override
    public NodeContentType getType() {
        return NodeContentType.FILE;
    }

    @Override
    public boolean equals(Object otherObject) {
        if (this == otherObject) {
            return true;
        }
        if (!(otherObject instanceof File)) {
            return false;
        }
        File that = (File) otherObject;
        return this.name.equals(that.name) &&
                (this.uri == null ? that.uri == null : this.uri.equals(that.uri));
    }

    @Override
    public int hashCode() {
        int result = name.hashCode();
        result = 31 * result + (uri != null ? uri.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return String.format(
                "File{name=%s, uri=%s}",
                this.name, this.uri);
    }
}

File: front-end/src/domain/value-object/file.ts
import { NodeContentType } from "../enumerate/node-content-type";
import { FileName } from "./file-name";
import { NodeContent } from "./node-content";

export class File extends NodeContent {
    private static readonly ERROR_FILENAME_NOT_DEFINED = "File name not defined";

    private readonly name: FileName;
    private readonly uri?: string; // optional

    constructor(name: FileName, uri?: string) {
        super();
        this.validateName(name);
        this.name = name;
        this.uri = uri;
    }

    getName(): FileName {
        return this.name;
    }

    getUri(): string | undefined {
        return this.uri;
    }

    hasUri(): boolean {
        return this.uri !== undefined && this.uri.trim() !== "";
    }

    setName(name: FileName): File {
        this.validateName(name);
        return new File(name, this.uri);
    }

    setUri(uri?: string): File {
        return new File(this.name, uri);
    }


    private validateName(name: FileName): void {
        if (!name) throw new Error(File.ERROR_FILENAME_NOT_DEFINED);
    }

    getType(): NodeContentType {
        return NodeContentType.FILE;
    }

    equals(other: unknown): boolean {
        if (!(other instanceof File)) return false;
        const otherFile = other as File;

        const nameEquals = this.name.equals(otherFile.getName());
        const uriEquals = this.uri === otherFile.getUri();

        return nameEquals && uriEquals;
    }

    toString(): string {
        return `File{name=${this.name.toString()}, uri=${this.uri ?? "undefined"}}`;
    }
}
